### ISSUE TITLE
Redis::CannotConnectError while starting the server

### ISSUE THREAD

**Issue by davidlin2k** [MarkUsProject/Markus#6471]
I have encountered an issue while trying to start the server with docker-compose. I have followed the exact procedure on the Wiki. The error shows Rack app ("GET /csc108/" - (172.28.0.1)): #<Redis::CannotConnectError: getaddrinfo_a: System error>.

However, it successfully starter the server for a few time but the error shows up again after I restarted the container. I was unable to reproduce the problem as it occasionally runs without issue.

Here's the full message for your reference. Any help is appreciated, thank you!

> markus-rails-1  | => Booting Puma
markus-rails-1  | => Rails 7.0.4.2 application starting in development 
markus-rails-1  | => Run `bin/rails server --help` for more startup options
markus-rails-1  | [44] Puma starting in cluster mode...
markus-rails-1  | [44] * Puma version: 6.1.0 (ruby 2.7.0-p0) ("The Way Up")
markus-rails-1  | [44] *  Min threads: 0
markus-rails-1  | [44] *  Max threads: 5
markus-rails-1  | [44] *  Environment: development
markus-rails-1  | [44] *   Master PID: 44
markus-rails-1  | [44] *      Workers: 3
markus-rails-1  | [44] *     Restarts: (✔) hot (✖) phased
markus-rails-1  | [44] * Preloading application
markus-rails-1  | [44] * Listening on http://0.0.0.0:3000
markus-rails-1  | [44] Use Ctrl-C to stop
markus-rails-1  | [54] WARNING hook before_worker_boot failed with exception (Redis::CannotConnectError) getaddrinfo_a: System error
markus-rails-1  | [44] - Worker 1 (PID: 54) booted in 2.0s, phase: 0
markus-rails-1  | [53] WARNING hook before_worker_boot failed with exception (Redis::CannotConnectError) getaddrinfo_a: System error
markus-rails-1  | [44] - Worker 0 (PID: 53) booted in 2.01s, phase: 0
markus-rails-1  | [56] WARNING hook before_worker_boot failed with exception (Redis::CannotConnectError) getaddrinfo_a: System error
markus-rails-1  | [44] - Worker 2 (PID: 56) booted in 2.0s, phase: 0

> Puma caught this error: getaddrinfo_a: System error (Redis::CannotConnectError)
/usr/lib/ruby/2.7.0/socket.rb:227:in `getaddrinfo'
/usr/lib/ruby/2.7.0/socket.rb:227:in `foreach'
/usr/lib/ruby/2.7.0/socket.rb:632:in `tcp'
/bundle/gems/redis-client-0.12.1/lib/redis_client/ruby_connection.rb:49:in `initialize'
/bundle/gems/redis-client-0.12.1/lib/redis_client.rb:664:in `new'
/bundle/gems/redis-client-0.12.1/lib/redis_client.rb:664:in `block in connect'
/bundle/gems/redis-client-0.12.1/lib/redis_client/middlewares.rb:12:in `connect'
/bundle/gems/redis-client-0.12.1/lib/redis_client.rb:663:in `connect'
/bundle/gems/redis-client-0.12.1/lib/redis_client.rb:657:in `raw_connection'
/bundle/gems/redis-client-0.12.1/lib/redis_client.rb:624:in `ensure_connected'
/bundle/gems/redis-client-0.12.1/lib/redis_client.rb:223:in `call_v'
/bundle/gems/redis-5.0.6/lib/redis/client.rb:73:in `call_v'
/bundle/gems/redis-5.0.6/lib/redis.rb:167:in `block in send_command'
/bundle/gems/redis-5.0.6/lib/redis.rb:166:in `synchronize'
/bundle/gems/redis-5.0.6/lib/redis.rb:166:in `send_command'
/bundle/gems/redis-5.0.6/lib/redis/commands/strings.rb:97:in `set'
/bundle/gems/redis-namespace-1.10.0/lib/redis/namespace.rb:558:in `wrapped_send'
/bundle/gems/redis-namespace-1.10.0/lib/redis/namespace.rb:515:in `call_with_namespace'
/bundle/gems/redis-namespace-1.10.0/lib/redis/namespace.rb:389:in `block (2 levels) in <class:Namespace>'
/bundle/gems/resque-2.4.0/lib/resque/data_store.rb:65:in `method_missing'
/bundle/gems/redis-namespace-1.10.0/lib/redis/namespace.rb:558:in `wrapped_send'
/bundle/gems/redis-namespace-1.10.0/lib/redis/namespace.rb:515:in `call_with_namespace'
/bundle/gems/redis-namespace-1.10.0/lib/redis/namespace.rb:389:in `block (2 levels) in <class:Namespace>'
/bundle/gems/rails_performance-1.0.4/lib/rails_performance/utils.rb:33:in `save_to_redis'
/bundle/gems/rails_performance-1.0.4/lib/rails_performance/models/request_record.rb:106:in `save'
/bundle/gems/rails_performance-1.0.4/lib/rails_performance/rails/middleware.rb:58:in `call!'
/bundle/gems/rails_performance-1.0.4/lib/rails_performance/rails/middleware.rb:38:in `call'
/bundle/gems/rails_performance-1.0.4/lib/rails_performance/rails/middleware.rb:17:in `call!'
/bundle/gems/rails_performance-1.0.4/lib/rails_performance/rails/middleware.rb:9:in `call'
/bundle/gems/actionpack-7.0.4.2/lib/action_dispatch/middleware/executor.rb:14:in `call'
/bundle/gems/actionpack-7.0.4.2/lib/action_dispatch/middleware/static.rb:23:in `call'
/bundle/gems/rack-2.2.6.2/lib/rack/sendfile.rb:110:in `call'
/bundle/gems/actionpack-7.0.4.2/lib/action_dispatch/middleware/host_authorization.rb:137:in `call'
/bundle/gems/rack-cors-2.0.0/lib/rack/cors.rb:102:in `call'
/bundle/gems/railties-7.0.4.2/lib/rails/engine.rb:530:in `call'
/bundle/gems/railties-7.0.4.2/lib/rails/railtie.rb:226:in `public_send'
/bundle/gems/railties-7.0.4.2/lib/rails/railtie.rb:226:in `method_missing'
/bundle/gems/rack-2.2.6.2/lib/rack/urlmap.rb:74:in `block in call'
/bundle/gems/rack-2.2.6.2/lib/rack/urlmap.rb:58:in `each'
/bundle/gems/rack-2.2.6.2/lib/rack/urlmap.rb:58:in `call'
/bundle/gems/puma-6.1.0/lib/puma/configuration.rb:269:in `call'
/bundle/gems/puma-6.1.0/lib/puma/request.rb:98:in `block in handle_request'
/bundle/gems/puma-6.1.0/lib/puma/thread_pool.rb:340:in `with_force_shutdown'
/bundle/gems/puma-6.1.0/lib/puma/request.rb:97:in `handle_request'
/bundle/gems/puma-6.1.0/lib/puma/server.rb:431:in `process_client'
/bundle/gems/puma-6.1.0/lib/puma/server.rb:233:in `block in run'
/bundle/gems/puma-6.1.0/lib/puma/thread_pool.rb:147:in `block in spawn_thread'

**Comment by david-yz-liu** [comment]
Thanks for reporting it, I've been experiencing this issue periodically (but not always; sometimes solved on restarting the containers). Will look into further!

**Comment by davidlin2k** [comment]
> Thanks for reporting it, I've been experiencing this issue periodically (but not always; sometimes solved on restarting the containers). Will look into further!

Thank you for confirming the issue! This seems like a major problem for me since I rarely could get it running. I have found that Redis gem with version >= 5.0.0 all causes this issue so it's where I suspect the root cause could be. By specifying a version in the Gemfile (gem 'redis', '~> 4.8.1' works for me), I was able to have the containers running without issues.
